File: src/main/java/june_2025/MinStack.java
package june_2025;


import java.util.Stack;

public class MinStack {

    Stack<Integer> stack;
    Stack<Integer> minStack;

    public MinStack() {
        stack = new Stack<>();
        minStack = new Stack<>();
    }

    public void push(int val) {
        stack.push(val);
        if (!minStack.isEmpty()) {
            int currMin = minStack.peek();
            minStack.push(Math.min(val, currMin));
        } else {
            minStack.push(val);
        }
    }

    public void pop() {
        if (stack.isEmpty()) {
            throw new RuntimeException("Stack is empty");
        }
        stack.pop();
        minStack.pop();
    }

    public int top() {
        if (stack.isEmpty()) {
            throw new RuntimeException("Stack is empty");
        }
        return stack.peek();
    }

    public int getMin() {
        if (minStack.isEmpty()) {
            throw new RuntimeException("Min Stack is empty");
        }
        return minStack.peek();
    }

}

File: python/min_stack.py
class MinStack(object):

    def __init__(self):
        self.stack = []
        self.minStack = []

    def push(self, val):
        """
        :type val: int
        :rtype: None
        """
        self.stack.append(val)
        if len(self.minStack) != 0:
            currMin = self.minStack[-1]
            if val < currMin:
                self.minStack.append(val)
            else:
                self.minStack.append(currMin)
        else:
            self.minStack.append(val)
        

    def pop(self):
        """
        :rtype: None
        """
        if len(self.stack) == 0:
            raise RuntimeError('Stack is empty')
        self.stack.pop()
        self.minStack.pop()
        

    def top(self):
        """
        :rtype: int
        """
        if len(self.stack) == 0:
            raise RuntimeError('Stack is empty')
        return self.stack[-1]
        

    def getMin(self):
        """
        :rtype: int
        """
        if len(self.minStack) == 0:
            raise RuntimeError('Min Stack is empty')
        return self.minStack[-1]


solution = MinStack()
solution.push(-2)
solution.push(0)
solution.push(-3)
print(solution.getMin())
solution.pop()
print(solution.top())
print(solution.getMin())